x/bank/module.go

```diff
@@ -0,0 +1,41 @@
+package bank
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	bankmodule "github.com/cosmos/cosmos-sdk/x/bank"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+	"github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	custombankkeeper "github.com/CudoVentures/cudos-node/x/bank/keeper"
+)
+
+type AppModule struct {
+	bankmodule.AppModule
+
+	keeper custombankkeeper.Keeper
+}
+
+// NewAppModule creates a new AppModule object
+func NewAppModule(cdc codec.Codec, keeper custombankkeeper.Keeper, accountKeeper types.AccountKeeper) AppModule {
+	bankModule := bankmodule.NewAppModule(cdc, keeper, accountKeeper)
+	return AppModule{
+		AppModule: bankModule,
+		keeper:    keeper,
+	}
+}
+
+// RegisterServices registers module services.
+// NOTE: Overriding this method as not doing so will cause a panic
+// when trying to force this custom keeper into a bankkeeper.BaseKeeper
+func (am AppModule) RegisterServices(cfg module.Configurator) {
+	types.RegisterMsgServer(cfg.MsgServer(), bankkeeper.NewMsgServerImpl(am.keeper))
```

Comment:
@nbaum, I'd like to clarify a crucial point that @tnv1 notified me, which causing the error when we directly use the method provided. Also the unit test is also catching it directly.

Custombankkeeper.Keeper is a struct, not an interface like the original bankkeeper. What the original RegisterServices is doing is that it use type assertion to pass the keeper to the NewMigrator method

![image](https://github.com/CudoVentures/cudos-node/assets/153589738/85939fb0-fcf0-4142-8af0-ecc843a19984)

In our code, we directly use the field, so no need for the type assertion at all

![image](https://github.com/CudoVentures/cudos-node/assets/153589738/2c80170a-80ab-4d90-a45a-8a5c91e73912)

Reply:
Ah! :bulb:

I should have tested that before commenting. :grin:

Then, an approach that _does_ work is passing the `keeper.BaseKeeper` to `NewAppModule`. The base app module then gets the `keeper.BaseKeeper` it expects, and `RegisterServices` doesn't need to be overridden.

```go
// NewAppModule creates a new AppModule object
func NewAppModule(cdc codec.Codec, keeper custombankkeeper.Keeper, accountKeeper types.AccountKeeper) AppModule {
	bankModule := bankmodule.NewAppModule(cdc, keeper.BaseKeeper, accountKeeper)
	return AppModule{
		AppModule: bankModule,
		keeper: keeper,
	}
}
```

This still works because the `bank` module itself never invokes BurnCoins on its keeper.

However, I'm happy with this as written.

This is really a bug with cosmos-sdk, isn't it? The bank's `AppModule` doesn't "really" work unless it gets a genuine `BaseKeeper`, but `NewAppModule` doesn't enforce that.